Test Command

Tyler-g-hudson opened the pull request:
Added the test command.

NOTE: This command works, but a large number of tests fail. I suspect that there is some issue with permissions inside of the docker image.

Tyler-g-hudson committed:
Updated _image.py to a new version

Tyler-g-hudson committed:
Added the "Test" command

Tyler-g-hudson committed:
Fixed most of the test failures, probably not a final fix

gmgunter commented on wigwam/_docker_cmake.py:
```diff
@@ -83,7 +83,11 @@ def cmake_build_dockerfile(base: str) -> str:
 
     # Build the project.
-    dockerfile += "RUN cmake --build $BUILD_PREFIX --parallel"
+    dockerfile += "RUN cmake --build $BUILD_PREFIX --parallel\n\n"
+
+    # Add permissions to the testing subdirectory under the build prefix.
+    # This step is necessary to enable testing on the image.
+    dockerfile += "RUN chmod -R 777 $BUILD_PREFIX"
```
```suggestion
    dockerfile += "RUN cmake --build $BUILD_PREFIX --parallel"

    # Add permissions to the testing subdirectory under the build prefix.
    # This step is necessary to enable testing on the image.
    dockerfile += "\n\nRUN chmod -R 777 $BUILD_PREFIX"
```

gmgunter commented on wigwam/_image.py:
```diff
@@ -1,10 +1,13 @@
 from subprocess import DEVNULL, CalledProcessError, run
 from sys import stdin
-from typing import Any, List, Optional, Type, TypeVar, Union, overload
+from typing import Any, Iterable, List, Optional, Type, TypeVar, overload
```
```suggestion
from collections.abc import Iterable
from typing import Any, List, Optional, Type, TypeVar, overload
```

gmgunter commented on wigwam/_image.py:
```diff
@@ -306,9 +321,9 @@ def drop_in(self, network: str = "host") -> None:
-            "bash", interactive=True, network=network, check=False
-        )  # pragma: no cover
+        self.run(  # pragma: no cover
+            "bash", interactive=True, network=network, check=False, host_user=True
```
Let's make this an option rather than hard-coding `host_user=True`

gmgunter commented on wigwam/commands.py:
```diff
@@ -252,6 +255,64 @@ def cmake_install(tag: str, base: str, no_cache: bool = False) -> Image:
 
+def test(
+    tag: str,
+    logfile: os.PathLike[str] | str,
```
```suggestion
    output_xml: os.PathLike[str] | str,
```

gmgunter commented on wigwam/commands.py:
```diff
@@ -252,6 +255,64 @@ def cmake_install(tag: str, base: str, no_cache: bool = False) -> Image:
+    compress_output : bool
+        If True, compress the output of the test.
+    quiet_fail : bool
+        If True, don't output on failure.
```
```suggestion
        If True, don't generate verbose output on failure.
```

gmgunter commented on wigwam/commands.py:
```diff
@@ -252,6 +255,64 @@ def cmake_install(tag: str, base: str, no_cache: bool = False) -> Image:
+
+    host_volume_path = "./Testing/Temps"
+    image_volume_path = "/tmp/Testing"
+    image: Image = Image(prefixed_tag)
```
```suggestion
    image = Image(prefixed_tag)
```

gmgunter commented on wigwam/commands.py:
```diff
@@ -252,6 +255,64 @@ def cmake_install(tag: str, base: str, no_cache: bool = False) -> Image:
+    image_volume_path = "/tmp/Testing"
+    image: Image = Image(prefixed_tag)
+
+    move_cmd = ["cd", str(build_prefix())]
```
```suggestion
    move_cmd = ["cd", os.fspath(build_prefix())]
```

gmgunter commented on wigwam/commands.py:
```diff
@@ -252,6 +255,64 @@ def cmake_install(tag: str, base: str, no_cache: bool = False) -> Image:
+    if not quiet_fail:
+        test_cmd += ["--output-on-failure"]
+
+    test_cmd += ["-T", "Test", "||", "true", ")"]
```
```suggestion
    # Note: as an alternative to copying the Test.xml file
    # from the default location to the specified output directory,
    # we could instead use `ctest --output-junit <file>`, although
    # this requires CMake>=3.21
    test_cmd += ["-T", "Test", "||", "true", ")"]
```

gmgunter commented on wigwam/commands.py:
```diff
@@ -252,6 +255,64 @@ def cmake_install(tag: str, base: str, no_cache: bool = False) -> Image:
+        f"{image_volume_path}/{logfile}",
+    ]
+
+    cmd: list[str] = move_cmd + ["&&"] + test_cmd + ["&&"] + file_cmd
```
```suggestion
    cmd = move_cmd + ["&&"] + test_cmd + ["&&"] + file_cmd
```

gmgunter commented on wigwam/commands.py:
```diff
@@ -252,6 +255,64 @@ def cmake_install(tag: str, base: str, no_cache: bool = False) -> Image:
+
+    cmd: list[str] = move_cmd + ["&&"] + test_cmd + ["&&"] + file_cmd
+
+    command = " ".join(cmd)
```
Give this a try and see if it causes any issues.

```suggestion
    command = shlex.join(cmd)
```

gmgunter commented on wigwam/commands.py:
```diff
@@ -252,6 +255,64 @@ def cmake_install(tag: str, base: str, no_cache: bool = False) -> Image:
+
+    prefixed_tag: str = prefix_image_tag(tag)
+
+    host_volume_path = "./Testing/Temps"
```
Instead of hard-coding this directory path, let's mount the parent directory of `logfile` in the container.

gmgunter commented on wigwam/commands.py:
```diff
@@ -252,6 +255,64 @@ def cmake_install(tag: str, base: str, no_cache: bool = False) -> Image:
+
+    command = " ".join(cmd)
+
+    host_vol_abspath = Path(host_volume_path).resolve()
+    host_vol_abspath.parent.mkdir(parents=True, exist_ok=True)
```
```suggestion
    host_vol_abspath = Path(logfile).parent.resolve()
    host_vol_abspath.mkdir(parents=True, exist_ok=True)
```

gmgunter commented on wigwam/commands.py:
```diff
@@ -252,6 +255,64 @@ def cmake_install(tag: str, base: str, no_cache: bool = False) -> Image:
+    host_vol_abspath.parent.mkdir(parents=True, exist_ok=True)
+
+    bind_mount = BindMount(
+        src=image_volume_path,
+        dst=host_vol_abspath,
```
Let's swap these two -- `src` should be the path on the host and `dst` should be the path inside the container.

```suggestion
        src=host_vol_abspath,
        dst=image_volume_path,
```

gmgunter commented on wigwam/cli/util_commands.py:
```diff
@@ -18,6 +18,26 @@ def init_util_parsers(subparsers: argparse._SubParsersAction, prefix: str) -> No
+        "tag", metavar="IMAGE_TAG", type=str, help="The tag or ID of the test image."
+    )
+    test_parser.add_argument(
+        "--logfile",
+        "-l",
```
```suggestion
        "--output-xml",
        "-o",
```

gmgunter commented on wigwam/cli/util_commands.py:
```diff
@@ -18,6 +18,26 @@ def init_util_parsers(subparsers: argparse._SubParsersAction, prefix: str) -> No
+        "-l",
+        type=str,
+        default="Test.xml",
+        help="The logfile to output test results to.",
```
```suggestion
        help="The output XML file to write test results to.",
```

Tyler-g-hudson committed:
Documentation fixes per PR

+ Exposed `default_user` to the CLI for `dropin` command
~ Changed `test` to mount to the parent of the given logfile instead of a hard-coded location
~ Changed `logfile` under `test` to `output_xml`
~ Changed some imports and type annotations to fit `__future__.annotations`

Tyler-g-hudson committed:
Updated lockfiles to avoid hdf5 1.14.2 dependency

Tyler-g-hudson committed:
Added the test test, fixed test, tested test (success!)

gmgunter commented on wigwam/commands.py:
```diff
@@ -252,19 +256,85 @@ def cmake_install(tag: str, base: str, no_cache: bool = False) -> Image:
+
+    xml_filename = Path(output_xml).name
+
+    image_volume_path = "/temporary/Testing"
```
Why did we change this from `/tmp/Testing` to `/temporary/Testing`?

Tyler-g-hudson replied:
I recall at some point having a discussion about `/tmp` being a directory that Linux does things with - I wanted to avoid any chance of the OS doing anything unusual with the directory while we're using it, so I moved over to `/temporary` which is just an arbitrary directory name.

gmgunter replied:
Hmm, the only particular special behavior that I'm aware of is that some OSes will clean up old files in `$TMPDIR` when the system boots.

I'm slightly wary of using `/temporary` instead of `/tmp` without a specific reason, just because I think it'd be confusing for folks reading the code since `/tmp` is the typical conventional location for temporary files.

If you prefer to avoid `/tmp`, how about calling it `/scratch` instead?

gmgunter reviewed (approved):
LGTM, nice work!

Tyler-g-hudson committed:
Testing container mount moved from `/temporary/Testing` to `/scratch/Testing`.
